### PR TITLE
:warning: Deprecate pkg/config/v1alpha1 types

### PR DESCRIPTION
**Pull request by vincepri**
This changeset sets deprecations and brings in changes so that settable controller options are now exported in the config.Controller struct. This package has been unmaintained for a while, and the inheritance logic makes it hard to rework the Options structs in the tree of controller runtime components.

Signed-off-by: Vince Prignano <vincepri@redhat.com>

### CODE REVIEW

File: pkg/config/controller.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	CacheSyncTimeout time.Duration
+
+	// RecoverPanic indicates whether the panic caused by reconcile should be recovered.
+	// Defaults to the Controller.RecoverPanic setting from the Manager if unset.
+	RecoverPanic *bool
+
+	// NeedLeaderElection indicates whether the controller needs to use leader election.
+	// Defaults to true, which means the controller will use leader election.
+	NeedLeaderElection *bool
+}
```
Comment (vincepri) [inline pkg/config/controller.go:49]: These options have to live in a different package, otherwise controller <-> manager get into a import cycle

**Comment by vincepri** [comment]
/retestOn Jan 27, 2023, at 11:55 AM, Kubernetes Prow Robot ***@***.***> wrote:﻿
@vincepri: The following test failed, say /retest to rerun all failed tests or /retest-required to rerun all mandatory failed tests:



Test name
Commit
Details
Required
Rerun command




pull-controller-runtime-test-master
cb5fcac
link
true
/test pull-controller-runtime-test-master



Full PR test history. Your PR dashboard. Please help us cut down on flakes by linking to an open issue when you hit one in your PR.

Instructions for interacting with me using PR comments are available here.  If you have questions or suggestions related to my behavior, please file an issue against the kubernetes/test-infra repository. I understand the commands that are listed here.

File: pkg/manager/manager.go [hunk 2]
```diff
@@ -37,7 +37,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/cluster"
 	"sigs.k8s.io/controller-runtime/pkg/config"
-	"sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/config/v1alpha1" //nolint:staticcheck // TODO: remove this import when v1alpha1 is removed
```
Comment (sbueringer) [inline pkg/manager/manager.go:40]: nit: this comment seems redundant, given that the code won't compile anymore once v1alpha1 is removed

File: pkg/manager/manager_options_test.go [hunk 3]
```diff
@@ -1,54 +0,0 @@
...
-	. "github.com/onsi/gomega"
-
-	"sigs.k8s.io/controller-runtime/pkg/config"
-
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
-	configv1alpha1 "sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
-)
-
-var _ = Describe("manager.Options", func() {
```
Comment (sbueringer) [inline pkg/manager/manager_options_test.go:14]: Shouldn't we keep the tests while the code is still there and only deprecated?
(similar in other cases)

**Review by sbueringer — COMMENTED** [review]
Overall lgtm

**Comment by sbueringer** [comment]
Didn't realize the restest merges it :0

**Comment by kerthcet** [comment]
Hi, controller-runtime users here, so if I depend on the configurations like `ControllerWebhook`, which is used to be part of the struct `v1alpha1.ControllerManagerConfigurationSpec`,  once we removed the v1alpha1, what I have to do is implement the `ControllerManagerConfigurationSpec` myself, anything can help to mitigate this? Thanks.

**Pull request merged**